File: js/gem-material.js
/**
 * gem-material.js
 *
 * Creates a physically-based gemstone material using THREE.MeshPhysicalMaterial
 * and provides presets for common gemstones.
 *
 * Each preset defines the optical properties that differentiate one stone
 * from another: IOR, dispersion, body color (attenuation), thickness, etc.
 * All presets share the same rendering flags (transmission, double-sided,
 * clearcoat) since the cut geometry is identical.
 */

import * as THREE from "three";

// ---------------------------------------------------------------
// Gemstone presets
//
// Each entry carries only the properties that vary per stone.
// Properties not listed here are shared defaults set in createGemMaterial().
// ---------------------------------------------------------------

/** @type {Object.<string, {label: string, props: Object}>} */
export const GEM_PRESETS = {
    diamond: {
        label: "Diamond",
        props: {
            ior: 2.42,
            dispersion: 0.044,
            thickness: 2.0,
            attenuationColor: new THREE.Color(0xffffff),
            attenuationDistance: 5.0,
            specularIntensity: 1.0,
            envMapIntensity: 2.5,
        },
    },
    sapphire: {
        label: "Blue Sapphire",
        props: {
            ior: 1.77,
            dispersion: 0.018,
            thickness: 2.5,
            attenuationColor: new THREE.Color(0.15, 0.18, 0.85),
            attenuationDistance: 2.0,
            specularIntensity: 0.9,
            envMapIntensity: 2.0,
        },
    },
    emerald: {
        label: "Green Emerald",
        props: {
            ior: 1.57,
            dispersion: 0.014,
            thickness: 2.2,
            attenuationColor: new THREE.Color(0.12, 0.72, 0.22),
            attenuationDistance: 1.8,
            specularIntensity: 0.85,
            envMapIntensity: 1.8,
        },
    },
    citrine: {
        label: "Amber Citrine",
        props: {
            ior: 1.55,
            dispersion: 0.013,
            thickness: 2.0,
            attenuationColor: new THREE.Color(0.92, 0.62, 0.10),
            attenuationDistance: 2.5,
            specularIntensity: 0.85,
            envMapIntensity: 2.0,
        },
    },
};

/** Ordered list of preset keys for cycling. */
export const GEM_ORDER = Object.keys(GEM_PRESETS);

// ---------------------------------------------------------------
// Material factory
// ---------------------------------------------------------------

/**
 * Create a gem material initialised to a given preset.
 *
 * @param {THREE.Texture|null} envMap
 * @param {string} [presetKey="diamond"]
 * @returns {THREE.MeshPhysicalMaterial}
 */
export function createGemMaterial(envMap = null, presetKey = "diamond") {
    const preset = GEM_PRESETS[presetKey] ?? GEM_PRESETS.diamond;

    const mat = new THREE.MeshPhysicalMaterial({
        // --- Shared across all gemstones ---
        transmission: 1.0,
        roughness: 0.0,
        metalness: 0.0,
        specularColor: new THREE.Color(0xffffff),
        clearcoat: 1.0,
        clearcoatRoughness: 0.0,
        envMap: envMap,
        side: THREE.DoubleSide,
        transparent: true,
        depthWrite: true,

        // --- Per-stone defaults ---
        ...preset.props,
    });

    return mat;
}

/**
 * Apply a preset to an existing material (in-place).
 * Only overwrites the properties defined in the preset — shared flags
 * (transmission, clearcoat, etc.) are left untouched.
 *
 * @param {THREE.MeshPhysicalMaterial} material
 * @param {string} presetKey
 */
export function applyGemPreset(material, presetKey) {
    const preset = GEM_PRESETS[presetKey] ?? GEM_PRESETS.diamond;
    for (const [key, value] of Object.entries(preset.props)) {
        if (material[key] instanceof THREE.Color) {
            material[key].copy(value);
        } else {
            material[key] = value;
        }
    }
    material.needsUpdate = true;
}

File: vulkan/src/scene/gem_material.c
/**
 * gem_material.c
 *
 * Gemstone preset definitions matching the web version exactly.
 */

#include "gem_material.h"

#include <string.h>

// ---------------------------------------------------------------------------
// Preset data (matching gem-material.js values)
// ---------------------------------------------------------------------------

static const char *PRESET_LABELS[GEM_PRESET_COUNT] = {
    "Diamond",
    "Blue Sapphire",
    "Green Emerald",
    "Amber Citrine",
};

static const GemMaterial PRESETS[GEM_PRESET_COUNT] = {
    [GEM_DIAMOND] = {
        .ior                = 2.42f,
        .dispersion         = 0.044f,
        .thickness          = 2.0f,
        .attenuationColor   = { 1.0f, 1.0f, 1.0f },
        .attenuationDistance = 5.0f,
        .specularIntensity  = 1.0f,
        .envMapIntensity    = 2.5f,
        .clearcoat          = 1.0f,
        .clearcoatRoughness = 0.0f,
        .roughness          = 0.0f,
        .metalness          = 0.0f,
    },
    [GEM_SAPPHIRE] = {
        .ior                = 1.77f,
        .dispersion         = 0.018f,
        .thickness          = 2.5f,
        .attenuationColor   = { 0.15f, 0.18f, 0.85f },
        .attenuationDistance = 2.0f,
        .specularIntensity  = 0.9f,
        .envMapIntensity    = 2.0f,
        .clearcoat          = 1.0f,
        .clearcoatRoughness = 0.0f,
        .roughness          = 0.0f,
        .metalness          = 0.0f,
    },
    [GEM_EMERALD] = {
        .ior                = 1.57f,
        .dispersion         = 0.014f,
        .thickness          = 2.2f,
        .attenuationColor   = { 0.12f, 0.72f, 0.22f },
        .attenuationDistance = 1.8f,
        .specularIntensity  = 0.85f,
        .envMapIntensity    = 1.8f,
        .clearcoat          = 1.0f,
        .clearcoatRoughness = 0.0f,
        .roughness          = 0.0f,
        .metalness          = 0.0f,
    },
    [GEM_CITRINE] = {
        .ior                = 1.55f,
        .dispersion         = 0.013f,
        .thickness          = 2.0f,
        .attenuationColor   = { 0.92f, 0.62f, 0.10f },
        .attenuationDistance = 2.5f,
        .specularIntensity  = 0.85f,
        .envMapIntensity    = 2.0f,
        .clearcoat          = 1.0f,
        .clearcoatRoughness = 0.0f,
        .roughness          = 0.0f,
        .metalness          = 0.0f,
    },
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

const char *gem_preset_label(GemPresetKey key)
{
    if (key < 0 || key >= GEM_PRESET_COUNT) return "Unknown";
    return PRESET_LABELS[key];
}

GemMaterial gem_preset_material(GemPresetKey key)
{
    if (key < 0 || key >= GEM_PRESET_COUNT) key = GEM_DIAMOND;
    return PRESETS[key];
}

GemPresetKey gem_preset_next(GemPresetKey current)
{
    return (GemPresetKey)(((int)current + 1) % GEM_PRESET_COUNT);
}

GemPresetKey gem_preset_prev(GemPresetKey current)
{
    return (GemPresetKey)(((int)current - 1 + GEM_PRESET_COUNT) % GEM_PRESET_COUNT);
}

GemMaterialUBO gem_material_to_ubo(const GemMaterial *mat)
{
    GemMaterialUBO ubo;
    memset(&ubo, 0, sizeof(ubo));

    ubo.ior                = mat->ior;
    ubo.dispersion         = mat->dispersion;
    ubo.thickness          = mat->thickness;
    ubo.attenuationDistance = mat->attenuationDistance;

    ubo.attenuationColor[0] = mat->attenuationColor[0];
    ubo.attenuationColor[1] = mat->attenuationColor[1];
    ubo.attenuationColor[2] = mat->attenuationColor[2];
    ubo.attenuationColor[3] = 1.0f;

    ubo.specularIntensity  = mat->specularIntensity;
    ubo.envMapIntensity    = mat->envMapIntensity;
    ubo.clearcoat          = mat->clearcoat;
    ubo.clearcoatRoughness = mat->clearcoatRoughness;

    ubo.roughness = mat->roughness;
    ubo.metalness = mat->metalness;

    return ubo;
}
